robotframework-ls/src/robotframework_ls/impl/flow_explorer_model_builder.py

```diff
@@ -1,9 +1,53 @@
+from contextlib import contextmanager
 from robotframework_ls.impl.protocols import ICompletionContext, IKeywordFound, INode
 from typing import List, Any, Dict, Tuple
 import os
 from robocorp_ls_core.basic import isinstance_name
 
 
+class _KeywordRecursionStack:
+    _stack: Dict
+
+    def __init__(self) -> None:
+        self._stack = {}
+
+    def has(self, keyword_name: str) -> bool:
+        return keyword_name in self._stack
+
+    def push(self, keyword_name: str) -> None:
+        self._stack.update({keyword_name: None})
+
+    def pop(self, keyword_name: str) -> None:
+        self._stack.pop(keyword_name)
+
+    @contextmanager
+    def scoped(self, keyword_name):
+        self.push(keyword_name)
+        yield
+        self.pop(keyword_name)
+
+
+class _UserKeywordCollector:
+    _user_keywords: List[Any]
```

Comment:
Typing could probably be `List[dict]`.

Sidenote (no need to do):

The actual keyword dict itself could be typed with a `TypedDict` so it could be `List[UserKeywordTypedDict]` (but then all the references to it would need to be typed too, not sure if it's worth the effort in this case).

See: https://peps.python.org/pep-0589/